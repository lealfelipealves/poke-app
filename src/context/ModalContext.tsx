import React, { createContext, useState, useContext } from "react";

type ModalContextProps = {
  children: React.ReactNode;
};

type ModalContextProviderProps = {
  isOpen?: boolean;
  setIsOpen: (open: boolean) => void;
};

export const ModalContext = createContext<ModalContextProviderProps>(
  {} as ModalContextProviderProps
);
ModalContext.displayName = "Modal Context";

export const ModalProvider = ({ children }: ModalContextProps) => {
  let [isOpen, setIsOpen] = useState(false);

  return (
    <ModalContext.Provider value={{ isOpen, setIsOpen }}>
      {children}
    </ModalContext.Provider>
  );
};

export const useModal = () => {
  const context = useContext(ModalContext);
  if (!context) {
    throw new Error(
      "Para usar o useModal, é obrigatório o usuário do Provider"
    );
  }
  return context;
};
