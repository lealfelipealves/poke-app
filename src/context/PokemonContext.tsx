import React, { createContext, useState, useMemo, useContext, useEffect } from 'react';
import { Pokemon, Regions } from 'pokenode-ts'
import { usePokemonInfiniteQuery } from '@/hooks/usePokemonInfiniteQuery';
import { NamedAPIResourceWithId } from '@/types';

type PokemonContextProps = {
  children: React.ReactNode
}

type PokemonContextProviderProps = {
  searchTerm?: string
  setSearchTerm: (searchTerm: string) => void,
  filteredData?: NamedAPIResourceWithId[],
  setFilteredData: (pokemonList: NamedAPIResourceWithId[]) => void,
  pokemon?: Pokemon
  setPokemon: (pokemon: Pokemon) => void,
  regionSelected?: Regions
  setRegionSelected: (region?: Regions) => void
}

export const PokemonContext = createContext<PokemonContextProviderProps>(
  {} as PokemonContextProviderProps
)
PokemonContext.displayName = 'Pokemon Context'

export const PokemonProvider = ({ children }: PokemonContextProps) => {
  const { data } = usePokemonInfiniteQuery();
  
  const [searchTerm, setSearchTerm] = useState("");
  const [filteredData, setFilteredData] = useState<NamedAPIResourceWithId[]>([]);
  const [pokemon, setPokemon] = useState<Pokemon>();  
  const [regionSelected, setRegionSelected] = useState<Regions | undefined>();

  const filterData = () => {
    const filtered = data?.pages.flatMap((page) =>
      page.results.filter((pokemon) => 
        pokemon.name.toLowerCase().includes(searchTerm.toLowerCase())
      )
    ).map((pokemon) => pokemon)
    setFilteredData(filtered ?? []);
  };
  
  useEffect(() => {
    filterData();
  }, [data, searchTerm]);

  const pokemonValue = useMemo(() => ({
    searchTerm, setSearchTerm,
    pokemon, setPokemon,
    filteredData, setFilteredData,
    regionSelected, setRegionSelected
  }), [
    searchTerm, setSearchTerm,
    pokemon, setPokemon,
    filteredData, setFilteredData,
    regionSelected, setRegionSelected
  ]);

  return (
    <PokemonContext.Provider value={pokemonValue}>
      {children}
    </PokemonContext.Provider>
  );
};

export const usePokemon = () => {
  const context = useContext(PokemonContext);
  if (!context) {
    throw new Error(
      'Para usar o usePokemon, é obrigatório o usuário do Provider'
    )
  }
  return context
}
