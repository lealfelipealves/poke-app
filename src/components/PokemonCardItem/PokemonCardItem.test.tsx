import { useModal } from '@/context/ModalContext';
import { usePokemon } from '@/context/PokemonContext';
import { fireEvent, render, screen } from '@testing-library/react';
import { PokemonCardItem } from './index';

const pokemon = {
  id: 1,
  name: "bulbasaur",
  url: "https://pokeapi.co/api/v2/pokemon/1/",
};

jest.mock('@/context/PokemonContext');
jest.mock('@/context/ModalContext');

describe('PokemonCardItem', () => {
  it('renders correctly', () => {
    const { container } = render(<PokemonCardItem pokemon={pokemon} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders pokemon name and id", () => {
    render(<PokemonCardItem pokemon={pokemon} />);
    const pokemonNameAndId = screen.getByText("#1 - bulbasaur");
    expect(pokemonNameAndId).toBeInTheDocument();
  });

});

describe('PokemonCardItem', () => {
  const pokemon = { id: 1, name: 'bulbasaur' };
  const setPokemon = jest.fn();
  const setIsOpen = jest.fn();
  usePokemon.mockReturnValue({ setPokemon, pokemon });
  useModal.mockReturnValue({ setIsOpen });

  beforeEach(() => {
    render(<PokemonCardItem pokemon={pokemon} />);
  });

  test('should call setPokemon and setIsOpen with the correct arguments when clicked', () => {
    fireEvent.click(screen.getByTestId('pokemon-card-item'));

    expect(setIsOpen).toHaveBeenCalledWith(true);
    expect(setPokemon).toHaveBeenCalledWith({
      ...pokemon,
      name: pokemon.name,
    });
  });
});