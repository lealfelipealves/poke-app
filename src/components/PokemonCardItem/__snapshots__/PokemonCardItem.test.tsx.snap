// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCardItem renders correctly 1`] = `
<div
  class="flex flex-col w-full rounded-2xl items-start justify-center bg-white font-bold text-lg p-4 cursor-pointer"
  data-testid="pokemon-card-item"
>
  <p>
    #
    1
     - 
    bulbasaur
  </p>
</div>
`;
